### PR TITLE
Applied memory leak fixes

### LINKED ISSUE #13: Memory leaks in the audio pipeline
From this [discussion](https://github.com/ILLIXR/ILLIXR/pull/112#issuecomment-837308821).

Changes found [here](https://github.com/ILLIXR/audio_pipeline/compare/4e7027b5baf91d5e29d73a3eff060491b10cf590...f600dec8bc531928ff9c72604e3ad45a4dc70891).

### PR DESCRIPTION
**Pull request by e3m3**
Closes #13.

### CODE REVIEW

File: src/audio.cpp [hunk 1]
```diff
@@ -57,38 +40,38 @@ void ILLIXR_AUDIO::ABAudio::loadSource(){
...
+            .fElevation = 3.14f/2,
+            .fDistance  = 1
+        });
+
+        soundSrcs.emplace_back("samples/radioMusicSample.wav", NORDER, true);
+        soundSrcs.back().setSrcPos({
+            .fAzimuth   = 1.0f, /// Should be same as lectureSample?
+            .fElevation = 0f,   /// Should be same as lectureSample?
+            .fDistance  = 5     /// Should be same as lectureSample?
+        });
```
Comment (e3m3) [inline src/audio.cpp]:
```suggestion
        soundSrcs.emplace_back("samples/radioMusicSample.wav", NORDER, true);
        soundSrcs.back().setSrcPos({
            .fAzimuth   = 1.0f,
            .fElevation = 0f,
            .fDistance  = 5
        });
```

**Review by charmoniumQ — APPROVED** [review]

File: include/sound.h [hunk 2]
```diff
@@ -1,42 +1,39 @@
...
+#pragma once
 
-#include <spatialaudio/Ambisonics.h>
 #include <string>
 #include <fstream>
+#include <memory>
+#include <spatialaudio/Ambisonics.h>
+
+constexpr std::size_t SAMPLERATE {48000U};
+constexpr std::size_t BLOCK_SIZE {512U};   /// Change to 1024U?
```
Comment (mhuzai) [inline include/sound.h]: 512 performs better on Intel's 32 KB data cache.

File: include/sound.h [hunk 3]
```diff
@@ -1,42 +1,39 @@
...
+constexpr std::size_t NORDER     {3U};
+constexpr std::size_t NUM_SRCS   {16U};    /// Change to 4U?
```
Comment (mhuzai) [inline include/sound.h]: 16 is good. It amortizes startup costs during profiling.

File: src/audio.cpp [hunk 4]
```diff
@@ -1,53 +1,36 @@
...
-    decoder = new CAmbisonicBinauralizer();
+    unsigned int tailLength {0U};
 
-    unsigned temp;
-    bool ok = decoder->Configure(NORDER, true, SAMPLERATE, BLOCK_SIZE, temp);
-    if (!ok) {
-        printf("Binauralizer Configuration failed!\n");
-    }
+    /// Binauralizer as ambisonics decoder
+    assert(decoder.Configure(NORDER, true, SAMPLERATE, BLOCK_SIZE, tailLength));
```
Comment (mhuzai) [inline src/audio.cpp]: I'm not sure if we should remove error checking from `opt` builds. Also, add failure message.

File: src/audio.cpp [hunk 5]
```diff
@@ -1,53 +1,36 @@
...
 #include <iostream>
 #include <cassert>
+#include "audio.h"
 
 
-ILLIXR_AUDIO::ABAudio::ABAudio(std::string outputFilePath, ProcessType procTypeIn) {
-    processType = procTypeIn;
+ILLIXR_AUDIO::ABAudio::ABAudio(std::string outputFilePath, ProcessType procTypeIn)
+    : ProcessType {procTypeIn}
+    , outputFile {
```
Comment (mhuzai) [inline src/audio.cpp:13]: Should this be closed in a destructor to make sure all the writes have been flushed to disk?

File: include/sound.h [hunk 6]
```diff
@@ -1,42 +1,39 @@
...
 
 namespace ILLIXR_AUDIO{
 	class Sound{
 	public:
-		Sound();
-		~Sound();
-
 		Sound(std::string srcFile, unsigned nOrder, bool b3D);
 		// set sound src position
-		void setSrcPos(PolarPoint& pos);
```
Comment (mhuzai) [inline include/sound.h:22]: Why make a copy?

**Review by mhuzai — COMMENTED** [review]
Since I can't comment on submodule diffs:
- I don't understand [this check](https://github.com/ILLIXR/libspatialaudio/compare/380107c4a6b0c26e20281b6f919092a9d8411c56...9bfd57756d8dd3c9194bdf931671b8265dc2b0d2#diff-97faad21c9f99856e1c6c19ca137c4dc11c65bf915fa2026c0f5ede03145c974R62). There is no pointer aliasing going on here, and the compiler/OS isn't going to allocate memory that overwrites a class member. So why is that check here?
- Why are there two of [these](https://github.com/ILLIXR/libspatialaudio/compare/380107c4a6b0c26e20281b6f919092a9d8411c56...9bfd57756d8dd3c9194bdf931671b8265dc2b0d2#diff-97faad21c9f99856e1c6c19ca137c4dc11c65bf915fa2026c0f5ede03145c974R73)?

File: src/audio.cpp (hunk 5 above)
Comment (e3m3) [reply src/audio.cpp:13]: [Looks like the destructor will take care of the flush](https://stackoverflow.com/questions/4802494/do-i-need-to-close-a-stdfstream).

File: include/sound.h (hunk 6 above)
Comment (e3m3) [reply include/sound.h:22]: Maybe there was a overwritten data in a previous implementation? Will change it back.

**Comment by e3m3** [comment]
>    I don't understand this check. There is no pointer aliasing going on here, and the compiler/OS isn't going to allocate memory that overwrites a class member. So why is that check here?

I agree. If a previous implementation used a static value for the buffer size then we would need it, but the `m_pfDelayBuffer` buffer is now heap allocated. Removing.

>    Why are there two of these?

If aliasing were to occur like the check above mentions, the `memset` could write a bogus value to `m_nDelayBufferLength`. Removing.

**Comment by e3m3** [comment]
Ready for review

**Review by mhuzai — APPROVED** [review]

**Pull request merged**